### DISCUSSION TITLE
Use external PostgreSQL database

### DISCUSSION THREAD

**Discussion by Mmoncadaisla** [developmentseed/k8s-eoapi#65] [Q&A] [answered]
### Context

>**NOTE:** Related to https://github.com/developmentseed/eoapi-k8s/discussions/64 and maybe to https://github.com/developmentseed/eoapi-k8s/issues/32 

The `eoAPI` helm chart allows enabling the `pgstac` service to spin up a PostgreSQL database, which is nice. 

However, for production-ready environments it'd be awesome if it was possible to directly pass the connection parameters to the `pgstac` user in an existent database for `titiler.pgstac`, `stac_fastapi.pgstac` and `tipg.pgstac` to use. 

In our case specifically, we aim to connect to a [Cloud Native PG](https://cloudnative-pg.io/) database, we understand that other users would benefit from connecting to Cloud SQL, Amazon RDS for PostgreSQL, etc too. 

### Question

Perhaps (hopefully) it is straightforward to achieve this and I missed the configuration steps to achieve it. Is it possible to connect to an external database instead of using the existent `db` service? 

If so, what are the requirements? we understand that a minimum version plus `pgstac` configured is a must (e.g: Postgresql>=13 and PostGIS>=3. with best performance  using PostGIS>=3.1 according to PgSTAC requirements).

**Comment by jrip-geo** [comment]
@Mmoncadaisla have you been able to figure this out, I'm interested in connecting the stac deployment to a cloud sql instance as well.

**Reply by Mmoncadaisla** [reply]
@jrip-geo not yet, I have had to push this for some time now, hopefully will get back to it in the coming weeks, I'll update this thread if there are any relevant findings

**Reply by MathewNWSH** [reply]
Hey @Mmoncadaisla,

any clue on how to launch PgSTAC with this [Cloud Natvie PG](https://cloudnative-pg.io/) manner? I'm about jump from my docker deployment to k8s as I look for production development. The goal is to serve the complete history of Sentinel missions. I fear that standard approach won't be enough.

Thanks,
Marc

**Reply by Mmoncadaisla** [reply]
@MathewNWSH I have not continued to explore the possibility to connect eoapi-k8s to an external PostgreSQL database yet. I think it'd be best if the team from developmentseed can take a look at this topic given that there's interest from others, perhaps @vincentsarago @ranchodeluxe can give a hand?

OTOH, if your goal is to leverage PgSTAC only then you can create the relations (schema, roles, tables and UDFs) via [PyPgSTAC](https://stac-utils.github.io/pgstac/pypgstac/) in any database, including a PostgreSQL database managed via Cloud Native PG operator.

**Reply by ranchodeluxe** [reply]
Sorry @Mmoncadaisla and @MathewNWSH: It's been brought to my attention that I'm missing discussion items (and also not getting alerted about them somehow 🥸). Thanks for having this discussion and using the helm chart.

This repository will probably be getting some love soon (though not sure by whom). As discussed soon [here](https://github.com/developmentseed/eoapi-k8s/issues/78) I think the goal will be to have the charts installing/using Crunchy's Postgres Operator by default for k8s-specific high availability solution

Setting up cloud provider storage access shouldn't be too hard but still needs to be documented. In terms of AWS (which I am more familiar with) the general steps would be:

- without IRSA (see last item for context of what that means) the pods should leverage the EC2's node IAM role. So you can add RDS policies grants to them like we do for other things such as [s3 here](https://github.com/developmentseed/eoapi-k8s-terraform/blob/main/terraform/aws/nodes.tf#L23-L37). This now enables every pod by default to talk to the cloud provider storage

- make sure each service optionally takes an "extraConfig" or "extraSecrets" param where folks can arbitrarily pass secret/configmap key/value pairs (or possibly just point to existing ones they've setup) to be mounted into the pods [like JH does](https://z2jh.jupyter.org/en/stable/administrator/advanced.html#hub-extraconfig). This would still have to be added to our helm configs

- decide if we want to leverage [ServiceRole IAM bridge for pods (IRSA)](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-enable-IAM.html) or not and add it to the Terraform optionally https://github.com/developmentseed/eoapi-k8s-terraform

In the future instead of creating a discussion for something like this I'd just create an issue and ask for thing you want! We also are open to PRs

**Reply by Mmoncadaisla** [reply]
Hi @ranchodeluxe thanks a lot for your attention!


>NOTE: perhaps the quick-access actions should be replaced then as it points to creating a new discussion for feature requests.
>
><img width="1203" alt="image" src="https://github.com/developmentseed/eoapi-k8s/assets/48254102/2fd7fcbb-4b4f-42de-86f7-e179b1bbf46d">


Per your comment it seems that eoapi-k8s could be more committed to choosing an operator for HA rather than allow users to configure an external PostgreSQL database (that might or might not leverage the Crunchy Data Operator) with some prerequisites. Should we go ahead and create an issue in this repo as a feature request then?

**Reply by ranchodeluxe** [reply]
> NOTE: perhaps the quick-access actions should be replaced then as it points to creating a new discussion for feature requests.

Good point 😉 

> Should we go ahead and create an issue in this repo as a feature request then?

There already is an existing issue created last week that can be fleshed out: https://github.com/developmentseed/eoapi-k8s/issues/77

**Reply by ranchodeluxe** [reply]
<img width="1188" alt="Screen Shot 2024-05-24 at 12 54 05 PM" src="https://github.com/developmentseed/eoapi-k8s/assets/208859/082cebdc-0505-461e-b88d-07c1ebb3b4e9">

Okay, at least the issue template is fixed now, thanks for the feedback

**Comment by pantierra** [comment, marked as answer]
This can be done by now and is [documented here](https://github.com/developmentseed/eoapi-k8s/blob/main/docs/configuration.md#database-configuration).
